Crear funcion de Lua para mostrar rama:hash de git

deivisCode opened the pull request:
O paquete gitinfo2 xoga demasiado cos Hooks de git e é un rollo. As solucions basadas en TeX puro...Están basadas en TeX puro (XD). Decidin facer de cero unha pequena funcion que mire no directorio .git e saque algúns datos relevantes, como o nome da rama actual e o hash.

Pode usarse en calquera parte con `\directlua{gitInfo()}`

IMPORTANTE:
- Engadín un macro de exemplo para mostralo na contraportada, é temporal, quitareino logo.
- A implementación desto é demencialmente básica. Non vos fiedes.
- Non funciona se tes un proxecto estructurado algo diferente. En concreto, se usas un 'bare repository' e varios worktrees (coma min XD) non funciona, dado que .git/HEAD e .git/refs/heads/[rama] non está onde deberían. Hei cambiar esto nalgún punto

En calqueira caso, esto é un bo exemplo de cómo usar Lua no proxecto

deivisCode committed:
Crear funcion de Lua para mostrar rama:hash de git

O paquete gitinfo2 xoga demasiado cos Hooks de git e é un rollo. As
solucions basadas en TeX puro...Están basadas en TeX puro (XD). Decidin
facer de cero unha pequena funcion que mire no directorio .git e saque
algúns datos relevantes, como o nome da rama actual e o hash.

IMPORTANTE:
- Engadín un macro de exemplo para mostralo na contraportada, é
  temporal, quitareino logo.
- A implementación desto é demencialmente básica. Non vos fiedes.
- Non funciona se tes un proxecto estructurado algo diferente. En
  concreto, se usas un 'bare repository' e varios worktrees (coma min
XD) non funciona, dado que .git/HEAD e .git/refs/heads/[rama] non está
onde deberían. Hei cambiar esto nalgún punto

En calqueira caso, esto é un bo exemplo de cómo usar Lua no proxecto

deivisCode committed:
Caso de traballar cun bare repo e worktrees

Tamen se engadiron monton de comentarios, e eliminei o exemplo na
contraportada

deivisCode commented:
Pola miña parte creo que xa estaría.

No proxecto de Evan Chen (Napkin) fixeime de que mostraban a versión exacta do proxecto no propio PDF. Pareceume unha idea curiosa, e pode que útil. Se temos varias versións dun PDF podemos asegurar de qué contribución concreta sairon. Atopara o paquete gitinfo2, o cal precisa que cambiemos varios hooks para xerar os datos que queremos, o cal non é moi complicado pero sí algo engorroso de facer cada vez que re-clonamos o proxecto. Ao final decidín facelo con Lua, abrindo directamente os arquivos que conteñen a info no repositorio e pista.

NOTAS:
- Esto é experimental, apenas fixen comprobacións para casos raros
- Inda que non furrulase, pode que a maior utilidade sexa presentar un exemplo de cómo se pode usar Lua directamente nun proxecto de LaTeX

De aprobalo, que se faga con *squash*, que neste caso queda máis limpo

arebla commented:
O de `\directlua{gitInfo()}` da os seguintes erros:
```tex
funcions.lua|41 error| Is a directory
funcions.lua|41 error| in function 'gitInfo'
[\directlua]|1 error| in main chunk. \directlua{gitInfo()}
```

Habería que facer algunha comprobación de que .git é efectivamente un ficheiro antes que de tente abrilo. Tamén metería `funcions.lua` dentro de  `trebellos/` ou `trebellos/lua/`

deivisCode commented:
> Habería que facer algunha comprobación de que .git é efectivamente un ficheiro antes que de tente abrilo

Logo de romperme a cabeza co tema resulta que existe LFS (non confundir con git LFS)

É posible facer
```lua
lfs.attributes(".git","mode")
```

o cal devolve `directory` ou `file`.

deivisCode committed:
(INACABADO) Intento de facer esto robusto

- Usar file:read(), que queda moitisimo mais limpo
- Comprobar se .git é file/directory con lfs.attributes
- Eliminar moitas variables liosas
- Facer todas as variables locais
- Volver engadir o exemplo na contraportada

Quedeime a medias porque me din conta de que a veces, o hash non se pode
atopar en $GIT_DIR/refs/heads/branch A veces, se hai moitas referencias,
ou usas 'git gc' ou 'maintenance', pode que se cree un arquivo
$GIT_DIR/packed-refs que ten todos os hashes e tal. Vese:
 ```
152204658ccc8c4126477370cdfb5de8b49348ab refs/heads/info_git
31623237990542b18db0abcc40dcb8d0d16434f9 refs/heads/principal
8fc61cadb3c1346d130455df27d1e921c53b728d refs/heads/snap
```
Hai que comprobar se existe ou no, e sacar o hash correctamente.

deivisCode commented:
Quedame o asunto dos hashes dentro de `.git/packed-refs`. O regex para sacar o hash neste caso e no caso normal é diferente, asique hai que meter máis `if` :(

deivisCode committed:
Rematado!!!

deivisCode committed:
Directorio para funcions de Lua

deivisCode commented:
Agora pode que sí que esté ben. Están algo caoticas as contribucions, e desactualizada a rama. Se ta todo ben fago outra limpa de cero mellor

deivisCode commented on lua/funcions.lua:
```diff
@@ -0,0 +1,182 @@
+
+            -- O ficheiro 'packed-refs' ten varias liñas como:
+            -- 152204658ccc8c4126477370cdfb5de8b49348ab refs/heads/[rama]
+            local _, _, hash = string.find( info , "^(%d+)%srefs/heads/"..rama .."$" )
```
o patrón debería ser `^(%x+)%srefs/heads/`, xa que o comezo son caracteres en hexadecimal. Inda así, parece que o find non atopa nada. Igual é problema de ler o arquivo con 'a'. Esto aplica ao outro bloque máis adiante

deivisCode replied:
Vale, parece que o problema é que ao ler con `read("a")`, o contido do arquivo pasa a ser unha soa cadena, cun so comezo `^` e sin `$`, osea xuntando todo. Hai que usar
`local hash = string.match( info , ".-(%x+)%srefs/heads/"..rama )`
`find` tamén vale pero `match` é mais correcto

deivisCode commented on lua/funcions.lua:
```diff
@@ -0,0 +1,182 @@
+            ficheiro_HASH:close()
+
+            local hash
+            _, _, hash = string.find( info , "^(%d+)%srefs/heads/"..rama .."$" )
```
Aquí o mesmo problema

deivisCode commented on lua/funcions.lua:
```diff
@@ -0,0 +1,182 @@
+            -- é problema, porque queremos a última referencia, a cal está en
+            -- .git/refs/heads/[rama], o cal é o PRIMEIRO sitio que buscamos.
+            -- Asique se existe, escollemos o correcto. Creo.
+            ficheiro_HASH = io.open(ruta_BASE .. "/packed-refs","r")
```
ruta_BASE debería ser simplemente ".git" neste caso, e non está definida. Hai que cambialo a `ficheiro_HASH = io.open(".git/packed-refs","r")`

deivisCode committed:
Rematado

- Patróns de regex correctos
- Usar .match en vez de .find
- Comentarios varios e links á documentacion
